internal/workflow_shadower.go

```diff
@@ -0,0 +1,305 @@
+		service  workflowserviceclient.Interface
+		options  *WorkflowShadowerOptions
+		replayer *WorkflowReplayer
+
+		status     int32
+		shutdownCh chan struct{}
+		shutdownWG sync.WaitGroup
+
+		clock clock.Clock
+	}
+)
+
+// NewWorkflowShadower creates an instance of the WorkflowShadower
+func NewWorkflowShadower(
+	service workflowserviceclient.Interface,
+	options *WorkflowShadowerOptions,
+) (*WorkflowShadower, error) {
+	if err := options.validateAndPopulateFields(); err != nil {
+		return nil, err
+	}
+	return &WorkflowShadower{
+		service:  service,
+		options:  options,
+		replayer: NewWorkflowReplayer(),
+
+		status:     statusInitialized,
+		shutdownCh: make(chan struct{}),
+
+		clock: clock.New(),
+	}, nil
+}
+
+// RegisterWorkflow registers workflow function to replay
+func (s *WorkflowShadower) RegisterWorkflow(w interface{}) {
+	s.replayer.RegisterWorkflow(w)
+}
+
+// RegisterWorkflowWithOptions registers workflow function with custom workflow name to replay
+func (s *WorkflowShadower) RegisterWorkflowWithOptions(w interface{}, options RegisterWorkflowOptions) {
+	s.replayer.RegisterWorkflowWithOptions(w, options)
+}
+
+// Run starts WorkflowShadower in a blocking fashion
+func (s *WorkflowShadower) Run() error {
+	if !atomic.CompareAndSwapInt32(&s.status, statusInitialized, statusStarted) {
+		return errors.New("Workflow shadower already started")
+	}
+
+	return s.shadowWorker()
+}
+
+// Stop stops WorkflowShadower and wait up to one miniute for all goroutines to finish before returning
+func (s *WorkflowShadower) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.status, statusStarted, statusStopped) {
+		return
+	}
+
+	close(s.shutdownCh)
+
+	if success := util.AwaitWaitGroup(&s.shutdownWG, time.Minute); !success {
+		s.options.Logger.Warn("Workflow Shadower timedout on shutdown")
+	}
+}
+
+func (s *WorkflowShadower) shadowWorker() error {
+	s.shutdownWG.Add(1)
+	defer s.shutdownWG.Done()
+
+	scanWorkflowReq := &shared.ListWorkflowExecutionsRequest{
+		Domain: common.StringPtr(s.options.Domain),
+		Query:  common.StringPtr(s.options.WorkflowQuery),
+	}
+	s.options.Logger.Info("Shadow workflow query",
+		zap.String("Query", s.options.WorkflowQuery),
+	)
+
+	ctx := context.Background()
+	expirationTime := time.Unix(0, math.MaxInt64)
+	if s.options.ExitCondition != nil && s.options.ExitCondition.ExpirationTime != 0 {
+		expirationTime = s.clock.Now().Add(s.options.ExitCondition.ExpirationTime)
+	}
+
+	replayCount := 0
+	maxReplayCount := math.MaxInt64
+	if s.options.ExitCondition != nil && s.options.ExitCondition.ShadowingCount != 0 {
+		maxReplayCount = s.options.ExitCondition.ShadowingCount
+	}
+	rand.Seed(s.clock.Now().UnixNano())
+	for {
+		scanWorkflowResp, err := s.scanWorkflowExecutionsWithRetry(ctx, scanWorkflowReq)
+		if err != nil {
+			return err
+		}
+
+		for _, execution := range scanWorkflowResp.Executions {
+			if s.clock.Now().After(expirationTime) {
+				return nil
+			}
+
+			if rand.Float64() >= s.options.SamplingRate {
```

Comment:
if the sampling is in the scanWorkflowExecutionsWithRetry, will it better if we want to convert it to an activity?

Reply:
Yes, I have already converted into in an activity in a later diff. But the activity diff is build on top of this one, so would like to land this one first.